### PR TITLE
Feature: Initial support for reading Standard SeedQR

### PR DESCRIPTION
**Pull request by kdmukai**
Adds SeedQR as an option when setting up a new "Bitcoin Core (hot wallet)" device.

Reads Standard SeedQRs:
* 12-word (25x25 QR)
* 24-word (29x29 QR)

See the [SeedQR Format Specification](https://github.com/SeedSigner/seedsigner/blob/dev/docs/seed_qr/README.md) for full details. It also includes example SeedQR images that you can use for testing.

Note that this also affects the "Add more keys" flow on an existing hot device (see last screenshot).

Includes placeholder code to eventually support the Compact SeedQR format. The Compact format is encoded as binary data that the js QR decoder can't currently interpret (it returns an empty string). I hope that we can upgrade the js QR decoder at some point but it only seems to exist in this repo as a minified js resource so it isn't really possible to make changes in there.

---

Additional changes:
* This PR reorders how the mnemonic is displayed. Rather than numbering across it will now number vertically: 4 columns for 24-word mnemonics; 2 columns for 12-word.
* The 2-column display could use some alignment tweaking but I didn't want to get bogged down in css.
* Some unnecessary instructions were removed.
* Warning message to back up the mnemonic moved to a slightly better position.

---

Screen Shot 2022-09-01 at 7 08 49 PM
Screen Shot 2022-09-01 at 7 09 42 PM
Screen Shot 2022-09-01 at 7 09 54 PM

---

The updated view when adding more keys to an existing hot device:

Screen Shot 2022-09-01 at 7 12 47 PM

### CODE REVIEW

**Comment by moneymanolis** [comment]
Failing test can be ignored, we are deprecating this test anyways.

**Comment by wombat6** [comment]
An observation: The "Bitcoin Core (hot wallet)" option under Add Device appears only after I've configured bitcoin core as my full node, which makes sense. But "Bitcoin Core (hot wallet)" _does_ appear in the dropdown when I choose Add Wallet > Manual Configuration, even if I haven't yet configured bitcoin core as my full node. Is that intentional?

**Comment by wombat6** [comment]
This PR worked for me. I imported the sample seed phrase using the QR in the SeedQR format specification. Seed phrase matched what was on the spec. Two wallets showed up in MyNode -- not sure if that's what's supposed to happen. See screenshot below:

Screen Shot 2022-09-05 at 8 09 15 PM

**Comment by wombat6** [comment]
Hmm, it looks like one of the wallets is associated with the device I created (which I called specter-seedqr-test), and the other is associated with a single sig wallet I created on top of that device (which I called specter_seedqr_test_wallet). I don't understand enough about how hot wallets work to know why there are two wallets on MyNode.

**Comment by moneymanolis** [comment]
> I don't understand enough about how hot wallets work to know why there are two wallets on bitcoin core.

There is the standard watch only wallet that Specter always uses and one hot wallet with private keys. Specter converts the mnemonic to an xprv which can be imported in Core as a descriptor (you can't import a mnemonic into Core, so Specter is bridging this).

**Comment by wombat6** [comment]
> > But "Bitcoin Core (hot wallet)" _does_ appear in the dropdown when I choose Add Device > Manual Configuration, even if I haven't yet configured bitcoin core as my full node. Is that intentional?
> 
> That should not be listed then I'd say. Can you create an issue for that @wombat6?

https://github.com/cryptoadvance/specter-desktop/issues/1875

File: src/cryptoadvance/specter/util/seedqr.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    # Parse the digitstream, 4 digits at a time
+    mnemonic = []
+    for i in range(0, int(len(digitstream) / 4)):
+        wordlist_index = int(digitstream[i * 4 : i * 4 + 4])
+        mnemonic.append(WORDLIST[wordlist_index])
+
+    return mnemonic
+
+
+def parse_compact_seedqr(bytestream: str, wordlist: List[str] = WORDLIST) -> List[str]:
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/util/seedqr.py:38]:
```suggestion
def parse_compact_seedqr(bytestream: str, wordlist: List[str] = WORDLIST) -> str:
```
Returns a string like so:
```

    bytestream = b'[\xbd\x9dq\xa8\xecy\x90\x83\x1a\xff5\x9dBeE'
    assert parse_compact_seedqr(bytestream) == "forum undo fragile fade shy sign arrest garment culture tube off merit"

```
See also https://embit.rocks/#/api/bip39?id=mnemonic_from_bytes:

_Returns
mnemonic string in the desired language._

File: src/cryptoadvance/specter/server_endpoints/wallets_api.py [hunk 2]
```diff
@@ -66,6 +67,22 @@ def generatemnemonic():
...
+@wallets_endpoint_api.route("/parse_seedqr/", methods=["POST"])
+@login_required
+def parse_seedqr():
+    digitstream = request.form.get("digitstream")
+    bytestream = request.form.get("bytestream")
+
+    if digitstream:
+        mnemonic = seedqr.parse_standard_seedqr(digitstream)
+    else:
+        mnemonic = seedqr.parse_compact_seedqr(bytestream)
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/server_endpoints/wallets_api.py:79]: `parse_compact_seedqr` already returns a string of the mnemonic - join should not be applied there.

File: src/cryptoadvance/specter/util/seedqr.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        logger.warn(f"Invalid Standard SeedQR datastream: {digitstream}")
+        return None
+
+    # Parse the digitstream, 4 digits at a time
+    mnemonic = []
+    for i in range(0, int(len(digitstream) / 4)):
+        wordlist_index = int(digitstream[i * 4 : i * 4 + 4])
+        mnemonic.append(WORDLIST[wordlist_index])
+
+    return mnemonic
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/util/seedqr.py:35]: Perhaps, return the result of `" ".join(mnemonic)` here to have it consistent with `parse_compact_seedqr`? That would also make the handling at `wallets_api.py` the same again.

File: src/cryptoadvance/specter/templates/device/new_device/new_device_mnemonic.jinja [hunk 4]
```diff
@@ -208,5 +303,66 @@
...
+                }
+
+                if (result.startsWith("StandardSeedQR:")) {
+                    let digitstream = result.substring("StandardSeedQR:".length);
+                    parseSeedQR(digitstream, null);
+
+                } else if (result.startsWith("CompactSeedQR:")) {
+                    let bytestream = result.substring("CompactSeedQR:".length);
+                    parseSeedQR(null, bytestream);
+                }
```
Comment (moneymanolis) [inline src/cryptoadvance/specter/templates/device/new_device/new_device_mnemonic.jinja:323]: Error feedback is missing here. I would at least add an else clause showing an error like "Sth. went wrong", perhaps with a reminder that compact SeedQR is not supported yet.

**Comment by moneymanolis** [comment]
https://github.com/cryptoadvance/specter-desktop/pull/1870/commits/5a17507b3ec6ed8279f49ab96c5fd91c7e1433ba handles unsupported QR code inputs with two error messages:

Scanning a Compact SeedQR, scanning a Standard SeedQR with unsupported length (say 18 word mnemonic) and all other QR codes result now in:

grafik

Scanning a Standard SeedQR with a wrong 4-digit (e.g. 3333) results in:

grafik

Pytests would need a little update (easily done) if you go with my suggestions for `parse_standard_seedqr()` @kdmukai.

**Comment by k9ert** [comment]
@kdmukai are you planning to work on this PR? Any comments on the remarks?

**Comment by kdmukai** [comment]
Getting caught up here again. I'll address @moneymanolis' points hopefully soon.

**Comment by k9ert** [comment]
>  I'll address @moneymanolis' points hopefully soon.

Hope you find some time here. I think we also need to make an extension out of this. I hope i can also find some time with this soon.

**Comment by moneymanolis** [comment]
Closing this after 6 months with no feedback from the author. Feel free to reopen.

**Pull request closed without merging**